consent/models.py

```diff
@@ -3,4 +3,150 @@
+        except DataSharingConsent.DoesNotExist:
+            return ProxyDataSharingConsent(**kwargs)
+
+
+class DataSharingConsentManager(models.Manager.from_queryset(DataSharingConsentQuerySet)):  # pylint: disable=no-member
+    """
+    Model manager for :class:`.DataSharingConsent` model.
+
+    Uses a QuerySet that returns a ``ProxyDataSharingConsent`` object when the searched-for
+    ``DataSharingConsent`` object does not exist. Otherwise behaves the same as a normal manager.
+    """
+
+    pass
+
+
+class ProxyDataSharingConsent(ConsentModelMixin):
+    """
+    A proxy-model of the ``DataSharingConsent`` model; it's not a real model, but roughly behaves like one.
+
+    Upon commit, a real ``DataSharingConsent`` object which mirrors the ``ProxyDataSharingConsent`` object's
+    pseudo-model-fields is created, returned, and saved in the database. The remnant, in-heap
+    ``ProxyDataSharingConsent`` object may be deleted afterwards, but if not, its ``exists`` fields remains ``True``
+    to indicate that the object has been committed.
+
+    NOTE: This class will be utilized when we implement program level consent by having an abstraction over these
+          consent objects per course.
+    """
+
+    objects = DataSharingConsentManager()
+
+    def __init__(self, enterprise_customer=None, username='', course_id='', granted=False):
+        """
+        Initialize a proxy version of ``DataSharingConsent`` which behaves similarly but does not exist in the DB.
+        """
+        self.enterprise_customer = enterprise_customer
+        self.username = username
+        self.course_id = course_id
+        self.granted = granted
+        self.exists = False
+
+    def commit(self):
+        """
+        Commit a real ``DataSharingConsent`` object to the database, mirroring current field settings.
+
+        :return: A ``DataSharingConsent`` object if validation is successful, otherwise ``None``.
+        """
+        try:
+            consent = DataSharingConsent.objects.create(
+                enterprise_customer=self.enterprise_customer,
+                username=self.username,
+                course_id=self.course_id,
+                granted=self.granted
+            )
+            self.exists = consent.exists
+            return consent
+        except (ValidationError, IntegrityError):
+            return None
+
+
+class Consent(TimeStampedModel):
+    """
+    An abstract base model for representing any type of consent.
+    """
+
+    class Meta:
+        """
+        Meta class for the ``Consent`` model.
+        """
+
+        abstract = True
+        app_label = 'consent'
+        unique_together = (("enterprise_customer", "username"),)
+
+    enterprise_customer = models.ForeignKey(
+        EnterpriseCustomer,
+        blank=False,
+        null=False,
+        related_name='enterprise_customer_consent',
+        on_delete=models.deletion.CASCADE
+    )
```

Comment:
Flexibility in the interest of being able to spin this into an IDA in future. Username is a unique indexed field on the Django user, so we don't lose much speed by not referencing the ID directly in the form of a ForeignKey.